제출자 최현준입니다.

che0223 opened the pull request:
좋은 강의 감사드립니다:)

che0223 committed:
1st commit

Moon-edu commented on tests/homework/csv_hw.py:
```diff
@@ -1,22 +1,38 @@
+    list_csv_data = []
+    header_col = None
+
+    with open('../../hw_data/assets.csv', encoding='utf-8') as f:
```
경로는 hw_data/assets.csv와 같이 hw_data/퍼일이름으로 해주세요

che0223 replied:
수정 완료했습니다

che0223 committed:
2nd commit

1. 경로 수정
../../hw_data/.csv -> hw_data/.csv

Moon-edu commented on tests/homework/csv_hw.py:
```diff
@@ -1,22 +1,38 @@
+            list_csv_data.append(data)
+
+    max_csv_data = max(list_csv_data, key=lambda csv_data: csv_data[header_col['est_asset_dollar']])
+    return (max_csv_data[header_col['name']], max_csv_data[header_col['est_asset_dollar']])
```
잘하셨습니다. 로직은 문제없어보이나 결과적으로는 오답이 나왔습니다.
```
def test_q1_test_find_richest_and_asset():
        result = find_richest_and_asset()
>       assert result == ("Xantha", 9977373)
E       AssertionError: assert ('Xantha', '9977373') == ('Xantha', 9977373)
E         At index 1 diff: '9977[37](https://github.com/Moon-edu/de-course-week4/actions/runs/7126292367/job/19413526068?pr=7#step:8:38)3' != 9977373
E         Full diff:
E         - ('Xantha', 9977373)
E         + ('Xantha', '9977373')
E         ?            +       +
```
한 가지 아쉬운 점이 있다면, 최대값을 찾기 위해 공간 복잡도 O(n) n|number of lines 를 써야할까요?
지금은 데이터가 적지만 파일 크기가 최소 수백MB에서 TB단위에서는 적절하지 않은 로직이 될 수도 있습니다.

Moon-edu commented on tests/homework/xml_hw.py:
```diff
@@ -1,22 +1,23 @@
+
+    min_est_asset_dollar_data = min(list_xml_data, key=lambda xml_data:xml_data['est_asset_dollar'])
 
+    return (min_est_asset_dollar_data['name'], min_est_asset_dollar_data['est_asset_dollar'])
```
```
def test_q4_test_find_poor_and_asset():
        result = find_poor_and_asset()
>       assert result == ('Noah', 28401)
E       AssertionError: assert ('Brandon', '1001611') == ('Noah', 28401)
E         At index 0 diff: 'Brandon' != 'Noah'
E         Full diff:
E         - ('Noah', 28401)
E         + ('Brandon', '1001611')
```

Moon-edu commented on tests/homework/csv_hw.py:
```diff
@@ -1,22 +1,38 @@
+        else:
+            dict_city[csv_data[header_col['city']]] = [int(csv_data[header_col['est_asset_dollar']])]
+
+    return [items[0] for items in sorted(dict_city.items(), key=lambda items: sum(items[1])/len(items[1]), reverse=True)[:3]]
```
잘하셨습니다. 나쁘지 않은 접근 방법입니다. 대략 복잡도를 계산해보면,
입력 데이터 O(n) - list_csv_data
dictionary O(n) - dict_city
sorted 에 의해 생성되는 새로운 list O(n) 
그래서 총 3 * number of line의 공간이 필요합니다.(똑같이 O(n)이긴합니다만...)
시간 복잡도 
최초 파일 읽기 - O(n)
dictionary 만들기 - O(n)
정렬 - O(nlogn)
결과적으로는 nlogn이긴한데 그래도 최대한 줄일 수 있는 것이 좋겠죠.
제 생각에는 O(1)의 공간복잡도와 O(n)의 시간 복잡도면 될 것 같습니다.
총 3개의 변수를 유지하거나 혹은 길이가 3인 Heap이라는 자료구조를 이용하면 가능할 것 같습니다.
구현을 강제하진 않습니다만, 한 번 고민해보는 것도 좋은 경험이 될 수 있을겁니다.

che0223 committed:
3rd commit

1. 로직 변경
1) csv_hw.py
1-1) find_richest_and_asset()
- 전체 csv 데이터 저장 부분 삭제(공간 낭비 최소화): 변수 하나로 max 데이터 검사 후 추출

1-2) find_top3_richest_city()
- 전체 데이터 list로 관리 -> dict로 city_name 값만 집계해서 효율적 관리(공간 낭비 최소화)
- 전체 데이터 sorted -> heapq 활용(시간 복잡도 최소화)

2) xml_hw.py
2-1) find_top3_poorest_city()
- 전체 데이터 list로 관리 -> dict로 city_name 값만 집계해서 효율적 관리(공간 낭비 최소화)
- 전체 데이터 sorted -> heapq 활용(시간 복잡도 최소화)

2. 버그 수정
1) xml.hw.py
- int() 형변환 추가

che0223 commented:
수정 완료했습니다:) 근데 궁금한 점이 생겼습니다.
csv경우에는 이터레이터 객체로 한줄 한줄 읽어서 메모리 상 문제가 없을 것 같은데,   json, xml, excel 경우에는 한번에 json.load(), xmltodict.parse(f.read())나 load_workbook()로 읽는 거 같아 파일 크기가 커질 경우 문제가 생길 것 같습니다, 그런 경우 해결책은 어떤게 있을까요?

제가 생각한 해결책으로는 json이나 xml은 텍스트 데이터라 f.read() 반복해서 읽는 방식으로 해결할 수 있을 것 같은데, excel 같은 바이너리 데이터는 잘 모르겠습니다;; 혹시 1바이트씩 읽으면서 처리해야되는걸까요?

Moon-edu commented:
좋은 질문입니다. 일반적으로 엄청나게 큰 json을 만드는 경우는 많지 않을 것 같긴하지만, 그래도 그런 고민을 하는 사람들이 있는 것 같습니다.
https://medium.com/@AlexanderObregon/json-streaming-how-to-work-with-large-json-files-efficiently-c7203de60ac2
streaming처리를 할 수 있게 해주는 라이브러리가 있습니다. 
xml도 있습니다.
https://pypi.org/project/xml-stream/
Excel
https://pypi.org/project/xlsx-streaming/

Moon-edu commented on tests/homework/xml_hw.py:
```diff
@@ -1,22 +1,30 @@
+    for city_name, city_data in dict_city.items():
+        heapq.heappush(list_heapq,((city_data[0] / city_data[1]), city_name))  # citydata(0: sum(est_asset_dollar), 1: count), list_heapq(0: sum(est_asset_dollar)/count, 1: city_name)
 
+    return [heapq.heappop(list_heapq)[1], heapq.heappop(list_heapq)[1], heapq.heappop(list_heapq)[1]]  # list_heapq(0: sum(est_asset_dollar)/count, 1: city_name)
```
```
    def test_q4_test_find_top3_poorest_city():
        result = find_top3_poorest_city()
        assert len(result) == 3
>       assert sorted(["Pinneberg", "Gorzow Wielkopolski", "Guri"]) == sorted(result)
E       AssertionError: assert ['Gorzow Wiel..., 'Pinneberg'] == ['Allistair', 'Medge', 'Noah']
E         At index 0 diff: 'Gorzow Wielkopolski' != 'Allistair'
E         Full diff:
E         - ['Allistair', 'Medge', 'Noah']
E         + ['Gorzow Wielkopolski', 'Guri', 'Pinneberg']
```

Moon-edu commented on tests/homework/csv_hw.py:
```diff
@@ -1,22 +1,42 @@
 def find_top3_richest_city() -> list:
-    pass
+    dict_city = {} # key: city, value: [sum(est_asset_dollar), count]
+    list_heapq = [] # 0: -sum(est_asset_dollar)/count, 1: city
```
한 가지 코멘트를 달자면 일단 적절한 코멘트를 다는 이 습관 매우 좋습니다.
하지만 동시에 아주 약간 아쉬운 점은, 좀 더 직관적인 코멘트가 좋지 않을까입니다.
대략 5초 정도 읽은 후에, 아 list_heapq 는 tuple을 담고 있구나라고 알아챘습니다.
그런데 주석이 조금 더 쉽게 쓰였으면 아마 읽자마자 바로 그렇구나 하고 넘어갔을 것 같습니다.

Moon-edu commented:
전반적으로 코드도 깔끔하고 알고리즘도 훌륭하게 구현하셔서 딱히 흠 잡기가 힘들었습니다. 그리고 streraming에 대한 질문까지 완벽했어요. LGTM 👍

che0223 committed:
4th commit

1. 버그 수정
1) xml_hw.py
find_top3_poorest_city()
'name' -> 'city'로 수정

che0223 commented:
피드백 감사합니다:D

che0223 committed:
5th commit

4th commit
